### PR TITLE
[8.x] Allow factory relationships to override state

### PR DESCRIPTION
**Pull request by imjohnbon**
**TLDR:** This PR allows factory relationships (like `for`) to override relationships generated within state.

Imagine you have a `PostFactory` in Laravel with a state called `withTopic` that automatically generates a topic for the post as well as some other post-specific topic related data:

```php
// inside PostFactory

public function withTopic()
{
    return $this->state(function () {
        return [
            'topic_id' => Topic::factory(),
            'topic_text_color' => $this->faker->color,
            'topic_background_color' => $this->faker->color,
        ];
    });
}
```

Now imagine inside a test you want to use the `withTopic` state, but override the topic that is generated with your own:

```php
// inside a test

$topic = Topic::factory()->create();
$post = Post::factory()->withTopic()->for($topic)->create();
```

Currently, in Laravel 8, the result of the code above will be **1 post created, 2 topics created**, and the topic associated with the post will be the one generated inside the state, not the one passed in via `for`.

With the change in this PR, the above code will now **create 1 post and 1 topic** (the one created in the test, not the state).

I believe this is more in line with what developers will expect when using factories. If a relationship is specifically passed in, it should override what is happening inside state methods.

### CODE REVIEW

**Comment by taylorotwell** [comment]
So, in your example, why didn't you make your `withTopic` method accept an optional `Topic` instance and handle the logic in there?

File: src/Illuminate/Database/Eloquent/Factories/Factory.php [hunk 1]
```diff
@@ -370,9 +370,9 @@ protected function getExpandedAttributes(?Model $parent)
     protected function getRawAttributes(?Model $parent)
     {
         return $this->states->pipe(function ($states) {
-            return $this->for->isEmpty() ? $states : new Collection(array_merge([function () {
+            return $this->for->isEmpty() ? $states : new Collection(array_merge($states->all(), [function () {
                 return $this->parentResolvers();
-            }], $states->all()));
+            }]));
```
Comment (taylorotwell) [inline src/Illuminate/Database/Eloquent/Factories/Factory.php:375]: A bit worried this is going to have more breaking change issues than we are anticipating.

**Comment by imjohnbon** [comment]
@taylorotwell I suppose I could have. To me, it was more a matter of expectation. I expected that if I passed in a specific model via `for` it would override whatever any state was doing. Similar to how in the old factories, if you did this:

```
factory(Post::class)->state('with_topic')->create(['topic_id' => $topic->id]);
```

The topic associated with your post would be `$topic`, not the one generated from state (I realize this isn't one to one in that regard).

Usually when I have an actual state method in a real factory, it's bigger than this fake example I made up. The state might modify up to 5-10 fields, with maybe 2 or 3 of those being relationships that create other models. I'd personally rather not have to potentially pass 2 or 3 optional arguments to a state method. I'd prefer to just call the state, and then override a relationship using `for` if I need to.

This is admittedly a bit of an edge case, but it's one I hit pretty quickly when converting my side project to the new factory system.

File: src/Illuminate/Database/Eloquent/Factories/Factory.php (hunk 1 above)
Comment (imjohnbon) [reply src/Illuminate/Database/Eloquent/Factories/Factory.php:375]: @taylorotwell I was worried about that as well, and you probably understand the potential implications more than I do. I will say this, I made this change on my side project which has 541 tests that all use the new factory system. This change didn't break any of the existing tests. Admittedly anecdotal, but it made me feel more confident about it.

**Comment by taylorotwell** [comment]
I think the breaking change of this is just a bit more than I'm willing to bite off right now. I still kinda prefer just passing the model into your state method.

**Comment by imjohnbon** [comment]
👍  Understandable, thanks.
